principles.md

```diff
@@ -0,0 +1,31 @@
+# usethis design principles
+
+*This is an experiment in making key package design principles explicit, versus only implicit in the code. The goal is to make maintenance easier, when spread out over time and across people.*
+
+## Active project
+
+Many usethis functions act on the **active project**, the path to which is stored in the internal environment `proj`, specifically in `proj$cur`. We do this instead of constantly passing around a base path. It is implied that such functions create or modify files inside the active project. This is mostly true of `use_*()` functions, though there are exceptions. For example, `use_course()` makes no reference to the active project.
+
+The project is activated upon first need, i.e. eventually some function calls `proj_get()` and, if `proj$cur` is `NULL`, we attempt to activate a project at (or above) current working directory.
+
+### Activation upon load or attach? No.
+
+We've contemplated project activation in `.onLoad()` or `.onAttach()`, but it's not clear which is more appropriate. Which suggests that neither is appropriate. If we ever do this, `zzz.R` would include something like this:
```

Comment:
Two more reasons not to activate on load/attach:

* If on attach, would mean that an package that imported usethis would also need to set on attach

* Current workflow allows you to open R, then `setwd()`, then usethis (not a workflow that we recommend, but one that is common in practice)